### ISSUE TITLE
Normals are not rendered correctly with CSGMesh

### ISSUE THREAD

**Issue by Toshiwoz** [godotengine/godot#24111]
I haven't found a related issue.

**Godot version:**
custom build 5b8dde4

**OS/device including version:**
Windows 10 integrated radeon graphic card.

**Issue description:**
![imagen](https://user-images.githubusercontent.com/21297356/48919936-cb03d880-ee63-11e8-9614-f3ea66493d92.png)
If I use an ArrayMesh I generate from code with CSGMesh, i get the normals as in the picture, quite strange as in a regular Mesh does not do that (see below).
![imagen](https://user-images.githubusercontent.com/21297356/48919984-0e5e4700-ee64-11e8-93ea-324b3ec0da4b.png)
It can be because the mesh is composed of 16 surfaces?
Or the fact that I generate the normals manually (might be some normals are not normalized)?
Not sure, there might be somethig wrong with my mesh array.

**Steps to reproduce:**
Create an empty project.
Use the mesh provided within a CSGMesh and compere with how it is rendered within MeshInstance.

**Minimal reproduction project:**
[terrain.zip](https://github.com/godotengine/godot/files/2609141/terrain.zip)

**Comment by BastiaanOlij** [comment]
Well I don't know what causes it (yet) but it behaves really weird on mine. I haven't looked at the normal generation code yet, might find some time during the week to dive into it.

**Comment by BastiaanOlij** [comment]
Finally had time to dive a little deeper. It would be nice to have a smaller mesh to play with but I think the root cause is due to many of the faces having (or getting) unique vertices. The strange looking triangles are all flat shaded which is the result of the logic recalculating normals not finding those vertices shared with other triangles.

I don't know if this is because of the way you've generated the mesh data or whether this is something Godot does when processing the data. Godot does tent to duplicate vertices when UVs or normals are not shared between the faces that vertice is used in.

I'm trying to see what Procmesh makes of it but it's not behaving atm :)

**Comment by Toshiwoz** [comment]
I can provide you the code that generates it:
https://github.com/Toshiwoz/GodotTerrainPrimitive

And here a smaller mesh, although it's not that small..
[terrain.zip](https://github.com/godotengine/godot/files/2717733/terrain.zip)

**Comment by Toshiwoz** [comment]
Hey, I guess it's better if I try to explain how I create the mesh.

I create the ArrayMesh by using a in index:

-  I added the vertices in the same order as the image.
- Then drawn the trianges using an index that draws 2 triangles per sqare, 1-7-6 and 1-2-7, then 2-3-7 and 3-8-7 and so on, alternatively
![image](https://user-images.githubusercontent.com/21297356/51351339-a8afc500-1a78-11e9-9316-a042bdee6510.png)

As I mentioned, if I display that as a regular MeshInstance, it works fine (I guess because it takes the normals I specify). It is just when I use the same in a CSGMesh that it does not work.

**Comment by Toshiwoz** [comment]
[terrain.zip](https://github.com/godotengine/godot/files/2770788/terrain.zip)

This is an even smaller example.

**Comment by Toshiwoz** [comment]
I also noticed another strange thing, when I repeatedly check and uncheck the "calculate tangents" property, it constantly changes, like it calculates it in a different way.

**Comment by BastiaanOlij** [comment]
Looks like that should work, I have to find time to dive deeper into this, its really weird. The only thing I can think of is that Godot ends up duplicating vertices.

When you check calculate tangents, does it then look right, or wrong?

**Comment by Toshiwoz** [comment]
> 
> When you check calculate tangents, does it then look right, or wrong?

Does still look wrong:

- Here the mesh is imported with calculate tangents disabled:
![image](https://user-images.githubusercontent.com/21297356/51762189-0d80a600-209d-11e9-96a2-8399dfa3270f.png)
- here after in cheched (by just checking nothing changes, it remains as the first picture) and uncheched that:
![image](https://user-images.githubusercontent.com/21297356/51762227-2be6a180-209d-11e9-88d4-41e4314ff609.png)

**Comment by Toshiwoz** [comment]
Now I changed the mesh to a capsule, then re-dragged the generated ArrayMesh and I get a different result:
![image](https://user-images.githubusercontent.com/21297356/51762361-8253e000-209d-11e9-9759-ac8b3a6ca75d.png)

I'm not sure if it's the snap value, I have changed that and the result changed again, when clearing and re-dragging the mesh:
![image](https://user-images.githubusercontent.com/21297356/51762454-c1823100-209d-11e9-8631-1a3498f45d76.png)

**Comment by Toshiwoz** [comment]
No, I think it's not the snap value.

So you say it's not the way I generate the array, in particular the array of normals?
Correct me if I'm wrong, the size of Vertex array is the same as Normals array and UVs right? The only one that can have a different size is the Vertex array, right?

I tried to generate with just one surface, and it looks better, but still, some triangles are not rendered right. Also I still experience some crashes due to the other issue I reported and you're working on (BTW, thanks for helping on that).

I hope this work will be useful for the coming native integration of terrain..

**Comment by BastiaanOlij** [comment]
I think the snap value only has an impact if you actually join two meshes together.

Yes, for every vertex you have a normal, then when the triangle is rendered the normal is interpolated over the three vertices and that result in the lighting. For a terrain like you build, each vertice only exists once which gives a smooth lighting over the whole surface of the terrain.

To get a flat shaded look, i.e. the triangles colour is near uniform, as in your screenshot you would need all normals of the three vertices of that triangle to point in roughly the same direction. That requires vertices to be duplicated because  you need a different normal for the same vertice to render each triangle properly.

Somewhere in the CSG logic that processes your mesh those vertices are duplicated. Or atleast, so is my theory :)

**Comment by BastiaanOlij** [comment]
Here is the culprit I think:
https://github.com/godotengine/godot/blob/master/modules/csg/csg_shape.cpp#L761

CSGMesh::_build_brush is the code that takes in the data from the mesh and converts it into what the CSG logic requires. At that link above you find a loop that loops through all the faces and will ensure that each face has unique vertices. If I read the code right the normals are actually discarded, instead each face is marked as either smooth or flat shaded by comparing how close the 3 normals are. On terrain, the flatter the terrain, the more likely the face is marked for flat shading resulting in the distinctly identifiable triangles in our output. 

When the CSG object is turned back into a mesh that information is used to recalculate the normals. I'll have a look if I can find a reason why its sometimes getting colors that do look a little off.

**Comment by BastiaanOlij** [comment]
Ok, ensuring that all faces are flat shaded gave the results I was expecting, but forcing all faces to be smooth shaded resulted in flat shaded polies strewn around the mesh.

I'm wondering if the hashmap logic is slightly too precise. In theory when the vertices where duplicated they became exact duplicates but if they're even slightly off, the hashmap logic won't recognise duplicates of a vertice are infact the same vertice...

**Comment by akien-mga** [comment]
Is this still reproducible in the current master branch or latest beta build?

**Comment by BastiaanOlij** [comment]
@akien-mga yes, it is still reproducible. As I wrote i think the solution of using a hashmap to determine if two vertices share the same physical space doesn't seem to work properly. Not sure why though and not sure how best to solve this.

**Comment by madmiraal** [comment]
I don't think this is a bug in Godot. I think, as @Toshiwoz originally suspected, it's the way the normals are calculated when the `ArrayMesh` is created.

`CSGMesh` actually ignores the `ArrayMesh` normals; unless all a face's vertices' normals are perpendicular:
https://github.com/godotengine/godot/blob/3b1c04550f9341856b1e94cda05a50eaebb3d35e/modules/csg/csg_shape.cpp#L843
If the normals are perpendicular it assumes that face should have a flat shader applied and not a smooth shader.
After completing the CSG operations, the `ArrayMesh` is reconstructed, and it calculates the vertex normals itself by averaging the normals of all the faces connected to each vertex:
https://github.com/godotengine/godot/blob/3b1c04550f9341856b1e94cda05a50eaebb3d35e/modules/csg/csg_shape.cpp#L301-L311
Depending on whether the face was previously identified as needing a flat or a smooth shader applied, the face's vertices' normals are either the normal of the face, or the average of the normals of the faces connected to that vertex:
https://github.com/godotengine/godot/blob/3b1c04550f9341856b1e94cda05a50eaebb3d35e/modules/csg/csg_shape.cpp#L380-L390

In the case of the `ArrayMesh` provided by @Toshiwoz, 5,083 of the 130050 faces (nearly 4%) have all the normals perpendicular, which are treated as requiring a flat shader applied, which results in the stepped appearance seen:
![CSGMeshFlatSurfaces](https://user-images.githubusercontent.com/9253928/77057436-dae5eb00-69cb-11ea-8f5c-7e52e5dac810.png)
If we ignore the normals provided and allow CSG to calculate all of them, we get the expected result:
![CSGMeshNoFlatSurfaces](https://user-images.githubusercontent.com/9253928/77057589-17194b80-69cc-11ea-8f23-08c7397f4694.png)

For completion, `MeshInstance` uses the vertex normals provided. If these were calculated using a different method (i.e. not the average of the normals of the faces connected to that vertex) it will look different:
![CSGMeshMeshInstance](https://user-images.githubusercontent.com/9253928/77058162-f7365780-69cc-11ea-9bff-cabe42851949.png)
Looking at @Toshiwoz's code, the normals are calculated using a different algorithm:
```
var pi_vector = Vector3(0.0,PI,0.0)
func calculate_normal_from_Heightsm(_mt_pxl, _y, _x, _py1, _px1) -> Vector3:
	var x = Vector3(
					atan(heights_array[_y - 1][_x - 1] - heights_array[_y + _py1][_x + _px1]),
					0.0,
					atan(heights_array[_y - 1][_x + _px1] - heights_array[_y + _py1][_x - 1])
					)
	var t = Vector3(
					atan(heights_array[_y][_x - 1] - heights_array[_y][_x + _px1]),
					0.0,
					atan(heights_array[_y - 1][_x] - heights_array[_y + _py1][_x])
					)
	return ((pi_vector+t+x)).normalized()
```

So the results are expected to be different.

Finally, regarding @BastiaanOlij comment on the hash algorithm, the `OAHashMap` converts the `Vector3` to a `String` and then uses `String`'s `hash()` function:
https://github.com/godotengine/godot/blob/af0b7f90bcdf1fca334932be2888d30aaef7bb75/core/ustring.cpp#L2227-L2236
However, this is largely irrelevant, because `OAHashMap` only uses this hash to identify the bin used to simplify the search for the key.
What is relevant is that when it calculates a vertex's normal by averaging the normals of the faces connected to that vertex, it currently only includes the normals from faces marked as `smooth`:
https://github.com/godotengine/godot/blob/3b1c04550f9341856b1e94cda05a50eaebb3d35e/modules/csg/csg_shape.cpp#L299
I think this is incorrect and will submit a PR to fix it.

In short, I think this issue can be closed.

**Comment by Calinou** [comment]
@madmiraal We might want to update the CSGMesh class documentation before closing this.

**Comment by akien-mga** [comment]
> We might want to update the CSGMesh class documentation before closing this.

That would be good to do now so that we can close the issue. Now we just need someone who understand what should go in that documentation :)
